internal/batches/executor/run_steps.go

```diff
@@ -23,6 +23,22 @@ import (
 	yamlv3 "gopkg.in/yaml.v3"
 )
 
+// stepExecutionResult is the executionResult after executing the step with the
+// given index in task.Steps.
+// TODO: The naming here kinda clashes with StepResult, which is used for
+// templating.
+type stepExecutionResult struct {
+	// Step is the index of the step in task.Steps
+	Step int `json:"step"`
```

Comment:
The type may be a bit more obvious with a slightly different name:

```suggestion
	StepIndex int `json:"step"`
```

I only suggest it because I feel like most people who see something called `Step` are going to immediately assume it's of the same type as its name, `Step`. 😅

Reply:
Nice! Thank you! I forgot about this. I even called this `StepIndex` in the `StepCacheKey` typ.